src/panels/lovelace/navigation/hui-lovelace-navigation.ts

```diff
@@ -0,0 +1,802 @@
+import { addSearchParam } from "../../../common/url/search-params";
+import { navigate } from "../../../common/navigate";
+import { showEditViewDialog } from "../editor/view-editor/show-edit-view-dialog";
+import { swapView } from "../editor/config-util";
+import { haStyle } from "../../../resources/styles";
+import "../../../components/ha-icon";
+import "../../../components/ha-icon-button";
+import "../../../components/ha-icon-button-arrow-next";
+import "../../../components/ha-icon-button-arrow-prev";
+import "../../../components/sl-tab-group";
+
+export interface NavigationConfig {
+  mode: "desktop" | "mobile";
+  editMode: boolean;
+  narrow: boolean;
+  curView?: number;
+  route?: {
+    path: string;
+    prefix: string;
+  };
+}
+
+@customElement("hui-lovelace-navigation")
+export class HuiLovelaceNavigation extends LitElement {
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @property({ attribute: false }) public lovelace?: Lovelace;
+
+  @property({ attribute: false }) public navigationConfig!: NavigationConfig;
+
+  @state() private _curView?: number | "hass-unused-entities";
+
+  protected updated(changedProperties: PropertyValues) {
+    super.updated(changedProperties);
+    if (changedProperties.has("navigationConfig")) {
+      this._curView = this.navigationConfig.curView;
+      this.setAttribute("mode", this.navigationConfig.mode);
+      if (this.navigationConfig.editMode) {
+        this.setAttribute("edit-mode", "");
+      } else {
+        this.removeAttribute("edit-mode");
+      }
+
+      if (
+        this.navigationConfig.mode === "mobile" &&
+        this.navigationConfig.editMode &&
+        typeof this._curView === "number"
+      ) {
+        this._scrollToActiveTab();
+      }
+    }
+  }
+
+  protected render(): TemplateResult | typeof nothing {
+    const views = this.lovelace?.config.views ?? [];
+
+    if (!this._shouldShowNavigation(views)) {
+      return nothing;
+    }
+
+    return this.navigationConfig.mode === "desktop"
+      ? this._renderDesktopTabs(views)
+      : this._renderMobileTabs(views);
+  }
+
+  private _shouldShowNavigation(views: LovelaceViewConfig[]): boolean {
+    const visibleViews = views.filter(
+      (view) => !view.subview && !this._isTabHiddenForUser(view)
+    );
+    return !(
+      this.navigationConfig.mode === "mobile" && visibleViews.length <= 1
+    );
+  }
+
+  private _isTabHiddenForUser(view: LovelaceViewConfig): boolean {
+    return (
+      view.visible !== undefined &&
+      ((Array.isArray(view.visible) &&
+        !view.visible.some((e) => e.user === this.hass!.user?.id)) ||
+        view.visible === false)
+    );
+  }
+
+  private _isViewHidden(view: LovelaceViewConfig): boolean {
+    return (
+      !this.navigationConfig.editMode &&
+      (view.subview || this._isTabHiddenForUser(view))
+    );
+  }
+
+  private _renderDesktopTabs(views: LovelaceViewConfig[]): TemplateResult {
```

Comment:
Since the rendering of desktop tabs and mobile tabs all have their own function, wouldn't it make more sense to split it into 2 components? It is a bit messy and hard to read now.